saw-core-coq/src/Verifier/SAW/Translation/Coq/Term.hs

```diff
@@ -443,85 +512,65 @@ mkDefinition name (Coq.Lambda bs t) (Coq.Pi bs' tp)
--- 'PiBinder' followed by zero or more implicit 'PiBinder's representing
--- additonal implicit typeclass arguments, added if the given type is @isort@,
--- @qsort@, etc.
-translatePiBinder ::
-  TermTranslationMonad m => (LocalName, Term) -> m [Coq.PiBinder]
-translatePiBinder (n, ty) =
-  translateBinder n ty >>= \case
-    (n',ty',[])
-      | n == "_"  -> return [Coq.PiBinder Nothing ty']
-      | otherwise -> return [Coq.PiBinder (Just n') ty']
-    (n',ty',nhs) ->
-      return $ Coq.PiBinder (Just n') ty' :
-               map (\(nh,nhty) -> Coq.PiImplicitBinder (Just nh) nhty) nhs
-
--- | Given a 'LocalName' and its type (as a 'Term'), return the translation of
--- the 'LocalName' as an 'Ident', the translation of the type as a 'Type',
--- and zero or more additional 'Ident's and 'Type's representing additonal
--- implicit typeclass arguments, added if the given type is @isort@, etc.
-translateBinder ::
-  TermTranslationMonad m =>
-  LocalName ->
-  Term ->
-  m (Coq.Ident,Coq.Type,[(Coq.Ident,Coq.Type)])
-translateBinder n ty@(asPiList -> (args, asSortWithFlags -> mb_sort)) =
+-- | The result of translating a SAW core variable binding to Coq, including the
+-- Coq identifier for the variable, the Coq translation of its type, and 0 or
+-- more implicit Coq arguments that apply to the variable
+data BindTrans = BindTrans { bindTransIdent :: Coq.Ident,
+                             bindTransType :: Coq.Type,
+                             bindTransImps :: [(Coq.Ident,Coq.Type)] }
+
+-- | Convert a 'BindTrans' to a list of Coq term-level binders
+bindTransToBinder :: BindTrans -> [Coq.Binder]
+bindTransToBinder (BindTrans {..}) =
+  Coq.Binder bindTransIdent (Just bindTransType) :
+  map (\(n,ty) -> Coq.ImplicitBinder n (Just ty)) bindTransImps
+
+-- | Convert a 'BindTrans' to a list of Coq type-level pi-abstraction binders
+bindTransToPiBinder :: BindTrans -> [Coq.PiBinder]
+bindTransToPiBinder (BindTrans { bindTransImps = [], .. })
+  | bindTransIdent == "_" = [Coq.PiBinder Nothing bindTransType]
+bindTransToPiBinder (BindTrans { bindTransImps = [], .. }) =
+  [Coq.PiBinder (Just bindTransIdent) bindTransType]
+bindTransToPiBinder (BindTrans{..}) =
+  Coq.PiBinder (Just bindTransIdent) bindTransType :
+  map (\(n,ty) -> Coq.PiImplicitBinder (Just n) ty) bindTransImps
+
+-- | Given a 'LocalName' and its type (as a 'Term'), translate the 'LocalName'
+-- to a Coq identifier, translate the type to a Coq term, and generate zero or
+-- more additional 'Ident's and 'Type's representing additonal implicit
+-- typeclass arguments, added if the given type is @isort@, etc. Pass all of
+-- this information to the supplied computation, in which the SAW core variable
+-- is bound to its Coq identifier.
+translateBinder :: TermTranslationMonad m => LocalName -> Term ->
+                   (BindTrans -> m a) -> m a
+translateBinder n ty@(asPiList -> (args, asSortWithFlags -> mb_sort)) f =
```

Comment:
`ty@(asPiList -> (args, asSortWithFlags -> mb_sort))` uses nested view patterns, which I find somewhat difficult to read. I'd propose turning one of these view patterns into a local `let` binding in the body of the function, especially since neither view pattern can fail to match.